src/psyclone/dynamo0p3.py

```diff
@@ -4269,6 +4316,117 @@ def _initialise_xoyoz_qr(self, parent):
+        quadrature.
+
+        :param parent: the node in the f2pygen AST representing the PSy \
+                       subroutine in which to insert the initialisation.
+        :type parent: :py:class:``psyclone.f2pygen.SubroutineGen`
+
+        '''
+        from psyclone.f2pygen import AssignGen, DeclGen
+
+        if "gh_quadrature_face" not in self._qr_vars:
+            return
+
+        qr_vars = ["np_xyz", "nfaces"]
+        qr_ptr_vars = ["weights_xyz"]
+        api_config = Config.get().api_conf("dynamo0.3")
+
+        for qr_arg_name in self._qr_vars["gh_quadrature_face"]:
+
+            # We generate unique names for the integers holding the numbers
+            # of quadrature points by appending the name of the quadrature
+            # argument
+            parent.add(
+                DeclGen(
+                    parent, datatype="integer",
+                    kind=api_config.default_kind["integer"],
+                    entity_decls=[name+"_"+qr_arg_name for name in qr_vars]))
+            decl_list = [name+"_"+qr_arg_name+"(:,:) => null()"
+                         for name in qr_ptr_vars]
+            parent.add(
+                DeclGen(parent, datatype="real", pointer=True,
+                        kind=api_config.default_kind["real"],
+                        entity_decls=decl_list))
+            # Get the quadrature proxy
+            proxy_name = qr_arg_name + "_proxy"
+            parent.add(
+                AssignGen(parent, lhs=proxy_name,
+                          rhs=qr_arg_name+"%"+"get_quadrature_proxy()"))
+            # Number of points in each dimension & no. of faces
+            for qr_var in qr_vars:
+                parent.add(
+                    AssignGen(parent, lhs=qr_var+"_"+qr_arg_name,
+                              rhs=proxy_name+"%"+qr_var))
+            # Pointers to the weights arrays
+            for qr_var in qr_ptr_vars:
+                parent.add(
+                    AssignGen(parent, pointer=True,
+                              lhs=qr_var+"_"+qr_arg_name,
+                              rhs=proxy_name+"%"+qr_var))
+
+    def _initialise_edge_qr(self, parent):
+        '''
+        Add in the initialisation of variables needed for edge
+        quadrature.
+
+        :param parent: the node in the AST representing the PSy subroutine
+                       in which to insert the initialisation
+        :type parent: :py:class:``psyclone.f2pygen.SubroutineGen`
+        '''
+        from psyclone.f2pygen import AssignGen, DeclGen
+
+        if "gh_quadrature_edge" not in self._qr_vars:
+            return
+
+        api_config = Config.get().api_conf("dynamo0.3")
+
+        for qr_arg_name in self._qr_vars["gh_quadrature_edge"]:
+            # We generate unique names for the integers holding the numbers
+            # of quadrature points by appending the name of the quadrature
+            # argument
+            decl_list = []
```

Comment:
That made me take a second look since the code here is (correctly) using the name-space manager and the other code wasn't. I've now reduced code duplication by having a single `_initialise_face_or_edge_qr` that can do either face or edge.

Reply:
This looks like a better approach, thanks. It does mean a "fresh" code review, though 😃